byzcoin/bcadmin/main.go

```diff
@@ -1062,6 +1069,95 @@ func darcShow(c *cli.Context) error {
 	return err
 }
 
+func debugFollow(c *cli.Context) error {
+	if c.NArg() < 1 {
+		return errors.New("please give the following arguments: url [bcID]")
+	}
+	if c.NArg() == 1 {
+		return debugList(c)
+	}
+
+	r := &onet.Roster{List: []*network.ServerIdentity{{URL: c.Args().First()}}}
+	if r == nil {
+		return errors.New("couldn't create roster")
+	}
+	bcID, err := hex.DecodeString(c.Args().Get(1))
+	if err != nil {
+		return err
+	}
+
+	local := onet.NewLocalTest(cothority.Suite)
+	defer local.CloseAll()
+	servers := local.GenServers(1)
```

Comment:
I tried first to instantiate the service without the whole context. But `onet.Context` has a lot of private fields, so I had to change all the `log.Lvl1(s.ServerIdentity(), ...`. And finally I thought that using a `LocalTest` was the easiest way to go. After some time it also felt normal, as this functionality _does_ some kind of local test.

> Two requests at the same time will fail because the port won't be open for the second.

I don't understand that. `LocalTest` uses channels, no? So why would it open a port?

Reply:
```go
// genLocalHosts returns n servers created with a localRouter
func (l *LocalTest) genLocalHosts(n int) []*Server {
	l.panicClosed()
	servers := make([]*Server, n)
	for i := 0; i < n; i++ {
		port := l.latestPort
		l.latestPort += 10
		servers[i] = l.NewServer(l.Suite, port)
	}
	return servers
}

// NewServer returns a new server which type is determined by the local mode:
// TCP or Local. If it's TCP, then an available port is used, otherwise, the
// port given in argument is used.
func (l *LocalTest) NewServer(s network.Suite, port int) *Server {
	l.panicClosed()
	var server *Server
	switch l.mode {
	case TCP:
		server = l.newTCPServer(s)
		// Set TLS certificate if any configuration available
		if len(l.webSocketTLSCertificate) > 0 && len(l.webSocketTLSCertificateKey) > 0 {
			cert, err := tls.X509KeyPair(l.webSocketTLSCertificate, l.webSocketTLSCertificateKey)
			if err != nil {
				panic(err)
			}
			server.WebSocket.Lock()
			server.WebSocket.TLSConfig = &tls.Config{Certificates: []tls.Certificate{cert}}
			server.WebSocket.Unlock()
		}
		server.StartInBackground()
	default:
		server = l.NewLocalServer(s, port)
	}
	return server
}
```

Reply:
Yes, but:

```
// NewLocalTest creates a new Local handler that can be used to test protocols
// locally
func NewLocalTest(s network.Suite) *LocalTest {
	dir, err := ioutil.TempDir("", "onet")
	if err != nil {
		log.Fatal("could not create temp directory: ", err)
	}

	return &LocalTest{
		Servers:    make(map[network.ServerIdentityID]*Server),
		Overlays:   make(map[network.ServerIdentityID]*Overlay),
		Services:   make(map[network.ServerIdentityID]map[ServiceID]Service),
		Trees:      make(map[TreeID]*Tree),
		Nodes:      make([]*TreeNodeInstance, 0, 1),
		Check:      CheckAll,
		mode:       Local,
		ctx:        network.NewLocalManager(),
		Suite:      s,
		path:       dir,
		latestPort: 2000,
	}
}
```

So it's in `Local` by default, and doing `NewLocalServer` - no?

Reply:
Oh god yes I missed that, nevermind then.